src/lib/support/Pool.h

```diff
@@ -0,0 +1,125 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *   Defines a memory pool class BitMapObjectPool.
+ */
+
+#pragma once
+
+#include <array>
+#include <assert.h>
+#include <atomic>
+#include <limits>
+#include <new>
+#include <stddef.h>
+
+namespace chip {
+
+class StaticAllocatorBase
+{
+public:
+    StaticAllocatorBase(size_t capacity) : mAllocated(0), mCapacity(capacity) {}
+
+    size_t Capacity(void) const { return mCapacity; }
+    size_t Allocated(void) const { return mAllocated; }
+    bool Exhausted(void) const { return mAllocated == mCapacity; }
+
+protected:
+    size_t mAllocated;
+    const size_t mCapacity;
+};
+
+class StaticAllocatorBitmap : public StaticAllocatorBase
+{
+protected:
+    /**
+     * Use the largest data type supported by `std::atomic`. Putting multiple atomic inside a single cache line won't improve
+     * concurrency, while the use of larger data type can improve the performance by reducing the number of outer loop iterations.
+     */
+    using tBitChunkType                         = unsigned long;
```

Comment:
Why `using` rather than `typedef`?

Reply:
It is personal preference. `using` is more modern and  can improve readability in some circumstances, so I prefer to use `using` than `typedef`.